marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -138,198 +147,427 @@ termToRange a { row, column } =
+      markSimplification constToVal SimplifiableValue b sb
+      pure (ValueSimp pos false t)
+
+lintObservation env t@(Term (ValueLE a b) pos) = do
+  sa <- lintValue env a
+  sb <- lintValue env b
+  case sa /\ sb of
+    (ConstantSimp _ _ c1 /\ ConstantSimp _ _ c2) -> pure (ConstantSimp pos true (c1 <= c2))
+    _ -> do
+      markSimplification constToVal SimplifiableValue a sa
+      markSimplification constToVal SimplifiableValue b sb
+      pure (ValueSimp pos false t)
+
+lintObservation env t@(Term (ValueEQ a b) pos) = do
+  sa <- lintValue env a
+  sb <- lintValue env b
+  case sa /\ sb of
+    (ConstantSimp _ _ c1 /\ ConstantSimp _ _ c2) -> pure (ConstantSimp pos true (c1 == c2))
+    _ -> do
+      markSimplification constToVal SimplifiableValue a sa
+      markSimplification constToVal SimplifiableValue b sb
+      pure (ValueSimp pos false t)
+
+lintObservation env t@(Term TrueObs pos) = do
+  pure (ConstantSimp pos false true)
+
+lintObservation env t@(Term FalseObs pos) = do
+  pure (ConstantSimp pos false false)
+
+lintObservation env hole@(Hole _ _ pos) = do
+  _ <- CMS.modify (over _holes (insertHole hole))
+  pure (ValueSimp pos false hole)
+
+lintValue :: LintEnv -> Term Value -> CMS.State State (TemporarySimplification BigInteger Value)
+lintValue env t@(Term (AvailableMoney acc token) pos) =
+  let
+    gatherHoles = getHoles acc <> getHoles token
   in
-    lintValue newState value
+    do
+      _ <- CMS.modify (over _holes gatherHoles)
+      pure (ValueSimp pos false t)
+
+lintValue env (Term (Constant (Term v pos2)) pos) = do
+  pure (ConstantSimp pos false v)
+
+lintValue env t@(Term (Constant h@(Hole _ _ _)) pos) = do
+  _ <- CMS.modify (over _holes (insertHole h))
+  pure (ValueSimp pos false t)
+
+lintValue env t@(Term (NegValue a) pos) = do
+  sa <- lintValue env a
+  case sa of
+    ConstantSimp _ _ c1 -> pure (ConstantSimp pos true (-c1))
+    _ -> do
+      markSimplification constToVal SimplifiableValue a sa
+      pure (ValueSimp pos false t)
+
+lintValue env t@(Term (AddValue a b) pos) = do
+  sa <- lintValue env a
+  sb <- lintValue env b
+  case sa /\ sb of
+    (ConstantSimp _ _ v1 /\ ConstantSimp _ _ v2) -> pure (ConstantSimp pos true (v1 + v2))
+    (ConstantSimp _ _ v /\ _) ->
+      if (v == zero) then
+        pure (simplifyTo sb pos)
+      else
+        defaultCase sa sb
+    (_ /\ ConstantSimp _ _ v) ->
+      if (v == zero) then
+        pure (simplifyTo sa pos)
+      else
+        defaultCase sa sb
+    _ -> defaultCase sa sb
+  where
+  defaultCase sa sb = do
+    markSimplification constToVal SimplifiableValue a sa
+    markSimplification constToVal SimplifiableValue b sb
+    pure (ValueSimp pos false t)
+
+lintValue env t@(Term (SubValue a b) pos) = do
+  sa <- lintValue env a
+  sb <- lintValue env b
+  case sa /\ sb of
+    (ConstantSimp _ _ v1 /\ ConstantSimp _ _ v2) -> pure (ConstantSimp pos true (v1 - v2))
+    (ConstantSimp _ _ v /\ _) ->
+      if (v == zero) then
```

Comment:
I think you can use guards in case expressions, might look nice here (and you've set the precedent by using them in functions already)

Reply:
Oh, yes, I didn't know you could do that, much better :)